controllers/componentversion_controller.go

```diff
@@ -261,70 +262,80 @@ func (r *ComponentVersionReconciler) reconcile(ctx context.Context, obj *v1alpha
 
 // parseReferences takes a list of references to embedded components and constructs a dependency tree out of them.
 func (r *ComponentVersionReconciler) parseReferences(ctx context.Context, parent *v1alpha1.ComponentVersion, references ocmdesc.References) ([]v1alpha1.Reference, error) {
-	log := log.FromContext(ctx)
 	result := make([]v1alpha1.Reference, 0)
 	for _, ref := range references {
-		// get component version
-		rcv, err := r.OCMClient.GetComponentVersion(ctx, parent, ref.ComponentName, ref.Version)
-		if err != nil {
-			return nil, fmt.Errorf("failed to get component version: %w", err)
-		}
-		// convert ComponentDescriptor to v3alpha1
-		dv := &compdesc.DescriptorVersion{}
-		cd, err := dv.ConvertFrom(rcv.GetDescriptor())
+		reference, err := r.parseReferencesLoopBody(ctx, parent, ref)
 		if err != nil {
-			return nil, fmt.Errorf("failed to convert component descriptor: %w", err)
-		}
-		// setup the component descriptor kubernetes resource
-		componentName, err := r.constructComponentName(ref.ComponentName, ref.Version, ref.GetMeta().ExtraIdentity)
-		if err != nil {
-			return nil, fmt.Errorf("failed to generate name: %w", err)
-		}
-		descriptor := &v1alpha1.ComponentDescriptor{
-			ObjectMeta: metav1.ObjectMeta{
-				Namespace: parent.GetNamespace(),
-				Name:      componentName,
-			},
-			Spec: v1alpha1.ComponentDescriptorSpec{
-				ComponentVersionSpec: cd.(*compdesc.ComponentDescriptor).Spec,
-				Version:              ref.Version,
-			},
+			return nil, fmt.Errorf("failed to parse reference: %w", err)
 		}
+		result = append(result, *reference)
+	}
+	return result, nil
+}
 
-		if err := controllerutil.SetOwnerReference(parent, descriptor, r.Scheme); err != nil {
-			return nil, fmt.Errorf("failed to set owner reference: %w", err)
-		}
+func (r *ComponentVersionReconciler) parseReferencesLoopBody(ctx context.Context, parent *v1alpha1.ComponentVersion, ref ocmdesc.ComponentReference) (*v1alpha1.Reference, error) {
```

Comment:
I think we should split this function because it does too many things. The naming also feels misleading to me because it does way more than parsing.

Reply:
I didn't want to refactor this in this PR. :D But you're right that it needs a bit of love. :)

Reply:
Done